uber/models.py

```diff
@@ -759,35 +766,43 @@ def auto_badge_num(self, badge_type):
             a specific range.
             :return:
             """
-            in_range = self.query(Attendee.badge_num).filter(Attendee.badge_num >= c.BADGE_RANGES[badge_type][0],
-                                                             Attendee.badge_num <= c.BADGE_RANGES[badge_type][1])
-            if in_range.count():
-                in_range_list = [int(row[0]) for row in in_range.order_by(Attendee.badge_num).all()]
+            in_range = self.query(Attendee.badge_num).filter(
+                Attendee.badge_num != None,
+                Attendee.badge_num >= c.BADGE_RANGES[badge_type][0],
+                Attendee.badge_num <= c.BADGE_RANGES[badge_type][1])
 
+            in_range_list = [int(row[0]) for row in in_range.order_by(Attendee.badge_num).all()]
+            if len(in_range_list):
                 # Searches badge range for a gap in badge numbers; if none found, returns the latest badge number + 1
                 # Doing this lets admins manually set high badge numbers without filling up the badge type's range.
                 start, end = c.BADGE_RANGES[badge_type][0], in_range_list[-1]
                 gap_nums = sorted(set(range(start, end + 1)).difference(in_range_list))
                 if not gap_nums:
-                    return in_range.order_by(Attendee.badge_num.desc()).first().badge_num + 1
+                    return end + 1
                 else:
                     return gap_nums[0]
             else:
                 return c.BADGE_RANGES[badge_type][0]
 
-        def shift_badges(self, badge_type, badge_num, *, until=None, **direction):
-            # assert_badge_locked()
-            until = until or c.BADGE_RANGES[badge_type][1]
+        def shift_badges(self, badge_type, badge_num, *, until=None, up=False, down=False):
             if not c.SHIFT_CUSTOM_BADGES or c.AFTER_PRINTED_BADGE_DEADLINE:
                 return False
-            assert not any(param for param in direction if param not in ['up', 'down']), 'unknown parameters'
-            assert len(direction) < 2, 'you cannot specify both up and down parameters'
-            down = (not direction['up']) if 'up' in direction else direction.get('down', True)
-            shift = -1 if down else 1
-            for a in self.query(Attendee).filter(Attendee.badge_num is not None,
-                                                 Attendee.badge_num >= badge_num,
-                                                 Attendee.badge_num <= until):
-                a.badge_num += shift
+
+            # assert_badge_locked()
+            from uber.badge_funcs import get_badge_type
+            (calculated_badge_type, error) = get_badge_type(badge_num)
```

Comment:
Whoops, accidentally posted that before I finished typing!

## Why do we want to call `get_badge_type()` at all?

I'm basically guarding against the case where an incorrect `badge_type` is passed as an argument.

Consider the case where a staff badge is being deleted, but for some reason `c.ATTENDEE_BADGE` is given as an argument for `badge_type`. Imagine the range for staff badges is 10-99 and the range for attendee badges is 100-999.

Badge Number | Badge Type | Attendee
---- | ---- | ----
10 | STAFF | Staffer A
11 | STAFF | Staffer B
12 | STAFF | Staffer C
13 | STAFF | Staffer D
... | ...
100 | ATTENDEE | Attendee A
101 | ATTENDEE | Attendee B
102 | ATTENDEE | Attendee C
103 | ATTENDEE | Attendee D

Now pretend we are deleting "Staffer B", only the arguments to `shift_badges()` are given as `badge_type=c.ATTENDEE` and `badge_number=11`. (_That's never supposed to happen, but these kinds of bugs happen all the time_) If we trust that `badge_type` is always correct and perform the badge shifting code based on that badge type, then `until`will be set to 999 and ALL the attendees will be shifted down:

Badge Number | Badge Type | Attendee
---- | ---- | ----
10 | STAFF | Staffer A
11 | STAFF | Staffer C
12 | STAFF | Staffer D
99 | ATTENDEE | Attendee A
... | ...
100 | ATTENDEE | Attendee B
101 | ATTENDEE | Attendee C
102 | ATTENDEE | Attendee D

Now we've got an attendee with a  badge number in the staff badge range. I'm not even sure what happens here. We may have some presave adjustments that convert `badge_type` to `c.STAFF_BADGE`. Or we may throw a validation error. I don't know, but it's definitely invalid, and not what we want.

By calculating the `badge_type` using `get_badge_type()`, we ensure that only the badges in the desired range are shifted.